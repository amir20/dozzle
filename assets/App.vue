<template lang="html">
  <main>
    <mobile-menu v-if="isMobile"></mobile-menu>
    <splitpanes @resized="onResize($event)">
      <pane min-size="10" :size="settings.menuWidth" v-if="!isMobile">
        <side-menu></side-menu>
      </pane>
      <pane :size="isMobile ? 100 : 100 - settings.menuWidth" min-size="10">
        <splitpanes>
          <pane class="has-min-height">
            <search></search>
            <router-view></router-view>
          </pane>
          <pane v-for="other in activeContainers" :key="other.id" v-if="!isMobile">
            <scrollable-view>
              <template v-slot:header>
                <container-title :value="other.name" closable @close="removeActiveContainer(other)"></container-title>
              </template>
              <log-viewer-with-source :id="other.id"></log-viewer-with-source>
            </scrollable-view>
          </pane>
        </splitpanes>
      </pane>
    </splitpanes>
  </main>
</template>

<script>
import { mapActions, mapGetters, mapState } from "vuex";
import { Splitpanes, Pane } from "splitpanes";

import LogViewerWithSource from "./components/LogViewerWithSource";
import ScrollableView from "./components/ScrollableView";
import SideMenu from "./components/SideMenu";
import MobileMenu from "./components/MobileMenu";
import Search from "./components/Search";
import ContainerTitle from "./components/ContainerTitle";

export default {
  name: "App",
  components: {
    LogViewerWithSource,
    SideMenu,
    MobileMenu,
    ScrollableView,
    Splitpanes,
    Pane,
    Search,
    ContainerTitle
  },
  data() {
    return {
      title: "",
      showNav: false
    };
  },
  metaInfo() {
    return {
      title: this.title,
      titleTemplate: "%s - Dozzle"
    };
  },
  async created() {
    await this.fetchContainerList();
    this.title = `${this.containers.length} containers`;
  },
  computed: {
    ...mapState(["containers", "activeContainers", "isMobile", "settings"])
  },
  methods: {
    ...mapActions({
      fetchContainerList: "FETCH_CONTAINERS",
      removeActiveContainer: "REMOVE_ACTIVE_CONTAINER",
      updateSetting: "UPDATE_SETTING"
    }),
    onResize(e) {
      if (e.length == 2) {
        this.updateSetting({ menuWidth: Math.min(90, e[0].size) });
      }
    }
  }
};
</script>

<style scoped lang="scss">
::v-deep .splitpanes__splitter {
  min-width: 4px;
  background: #666;
  &:hover {
    background: rgb(255, 221, 87);
  }
}

.button.has-no-border {
  border-color: transparent !important;
}

.has-min-height {
  min-height: 100vh;
}
</style>
