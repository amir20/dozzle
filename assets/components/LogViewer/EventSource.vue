<template>
  <InfiniteLoader :onLoadMore="fetchMore" :enabled="messages.length > 50"></InfiniteLoader>
  <slot :messages="messages"></slot>
</template>

<script lang="ts" setup generic="T">
import { LogStreamSource } from "@/composable/eventStreams";

const loadingMore = defineEmit<[value: boolean]>();

const props = defineProps<{
  streamSource: (t: Ref<T>) => LogStreamSource;
  entity: T;
}>();

const { entity, streamSource } = toRefs(props);

const { messages, loadOlderLogs } = streamSource.value(entity);

const beforeLoading = () => loadingMore(true);
const afterLoading = () => loadingMore(false);

defineExpose({
  clear: () => (messages.value = []),
});

const fetchMore = () => loadOlderLogs({ beforeLoading, afterLoading });
</script>
