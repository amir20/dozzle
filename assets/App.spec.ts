import EventSource from "eventsourcemock";
import { shallowMount, RouterLinkStub } from "@vue/test-utils";
import { createStore, Store } from "vuex";
import App from "./App.vue";

jest.mock("./store/config.ts", () => ({ base: "" }));

jest.mock("~icons/octicon/download-24", () => {}, { virtual: true });
jest.mock("~icons/octicon/trash-24", () => {}, { virtual: true });
jest.mock("~icons/mdi-light/chevron-double-down", () => {}, { virtual: true });
jest.mock("~icons/mdi-light/chevron-left", () => {}, { virtual: true });
jest.mock("~icons/mdi-light/chevron-right", () => {}, { virtual: true });
jest.mock("~icons/mdi-light/magnify", () => {}, { virtual: true });
jest.mock("~icons/cil/columns", () => {}, { virtual: true });
jest.mock("~icons/octicon/container-24", () => {}, { virtual: true });
jest.mock("~icons/mdi-light/cog", () => {}, { virtual: true });

describe("<App />", () => {
  const stubs = {
    RouterLink: RouterLinkStub,
    "router-view": true,
    "chevron-left-icon": true,
    "chevron-right-icon": true,
  };

  let store: Store<State>;

  interface State {
    settings: { menuWidth: number };
    containers: { id: string; name: string }[];
  }

  beforeEach(() => {
    global.EventSource = EventSource;
    const state = {
      settings: { menuWidth: 15 },
      containers: [{ id: "abc", name: "Test 1" }],
    };

    const getters = {
      visibleContainers(store: State): { id: string; name: string }[] {
        return store.containers;
      },
      activeContainers() {
        return [];
      },
    };

    store = createStore({ state, getters });
  });

  test("has right title", async () => {
    const wrapper = shallowMount(App, {
      global: {
        plugins: [store],
        stubs,
      },
    });
    wrapper.vm.$store.state.containers = [
      { id: "abc", name: "Test 1" },
      { id: "xyz", name: "Test 2" },
    ];
    await wrapper.vm.$nextTick();

    expect(wrapper.vm.title).toContain("2 containers");
  });

  test("renders correctly", async () => {
    const wrapper = shallowMount(App, {
      global: {
        plugins: [store],
        stubs,
      },
    });
    await wrapper.vm.$nextTick();
    expect(wrapper.html()).toMatchSnapshot();
  });
});
