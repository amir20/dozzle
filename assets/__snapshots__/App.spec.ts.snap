// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> renders correctly 1`] = `
"<main>
  <!--v-if-->
  <splitpanes-stub horizontal=\\"false\\" pushotherpanes=\\"true\\" dblclicksplitter=\\"true\\" rtl=\\"false\\" firstsplitter=\\"false\\"></splitpanes-stub><button class=\\"button is-rounded is-settings-control\\" id=\\"hide-nav\\"><span class=\\"icon\\"><mdi-light-chevron-left></mdi-light-chevron-left></span></button>
</main>"
`;
